### ISSUE TITLE
pattern for enabling/disabling UI components

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/sun#241]
This came up in https://github.com/phetsims/scenery-phet/issues/229 (NumberControl should have enabled/disable). PhET UI components (primarily in sun and scenery-phet) currently implement a variety of patterns for enabling and disabling those components.  And it's not clear which pattern to use in NumberControl, or components in general.

This issue considers:

• What are the current patterns?
• Are there other patterns that should be considered?
• Should PhET standardize on one pattern, or is variation justified?
• Is this topic applicable only to UI components, or does it apply to other things (e.g. model elements) that can be enabled/disabled?

**Comment by pixelzoom** [comment]
# Pattern #1:  Constructor option

In this pattern, an `enabledProperty` of type `{Property.<boolean>}` may be provided as an option to the UI component's constructor.  The client is free to ignore this option if there's no need to disable the component. The component links to `enabledProperty` to manage its state, and must unlink from `enabledProperty` in `dispose`.

Typical implementation:

```js
/**
 * @param {Property.<boolean>} enabledProperty
 * @constructor
 */
function Component( enabledProperty, …. ) {

   var enabledObserver = function( enabled ) {
     // enable or disable the component
   };
   enabledProperty.link( enabledObserver );

   // @private
   this.disposeComponent = function() {
     enabledProperty.unlink( enabledObserver );
   }
}

// prototype properties
return inherit( Supertype, Component, {

  // @public
  dispose: function() { this.disposeComponent(); }
} );
```

Usage 1.1:

```js
var componentEnabledProperty = new Property( true );
var component = new Component( componentEnabledProperty, … );
var checkBox = new CheckBox( label, componentEnabledProperty );

// cleanup
component.dispose();
checkBox.dispose();
```

Usage 1.2:

```js
var componentEnabledProperty = new Property( true );
var component = new Component( componentEnabledProperty, … );

var runningProperty = new Property( true ); // is the sim running?
var runningObserver = function( running ) {
  componentEnabledProperty.set( !running ); // disable component when sim is running
  // … and anything else that needs to change
};
runningProperty.link( runningObserver );

var checkBox = new CheckBox( label, runningProperty );

// cleanup
component.dispose();
runningProperty.unlink( runningObserver ); 
checkBox.dispose();
```

**Comment by pixelzoom** [comment]
# Pattern #1:  Require constructor parameter

In this pattern, an `enabledProperty` of type `{Property.<boolean>}` must be provided as an argument to the UI component's constructor.  The client must provide this Property even if it has no need to disable the component. The component links to `enabledProperty` to manage its state, and must unlink from `enabledProperty` in `dispose`.

Typical implementation:

```js
/**
 * @param {Property.<boolean>} enabledProperty
 * @constructor
 */
function Component( enabledProperty, …. ) {

   var enabledObserver = function( enabled ) {
     // enable or disable the component
   };
   enabledProperty.link( enabledObserver );

   // @private
   this.disposeComponent = function() {
     enabledProperty.unlink( enabledObserver );
   }
}

// prototype properties
return inherit( Supertype, Component, {

  // @public
  dispose: function() { this.disposeComponent(); }
} );
```

Usage 1.1:

```js
var componentEnabledProperty = new Property( true );
var component = new Component( componentEnabledProperty, … );
var checkBox = new CheckBox( label, componentEnabledProperty );

// cleanup
component.dispose();
checkBox.dispose();
```

Usage 1.2:

```js
var componentEnabledProperty = new Property( true );
var component = new Component( componentEnabledProperty, … );

var runningProperty = new Property( true ); // is the sim running?
var runningObserver = function( running ) {
  componentEnabledProperty.set( !running ); // disable component when sim is running
  // … and anything else that needs to change
};
runningProperty.link( runningObserver );

var checkBox = new CheckBox( label, runningProperty );

// cleanup
component.dispose();
runningProperty.unlink( runningObserver ); 
checkBox.dispose();
```

**Comment by pixelzoom** [comment]
# Pattern 3: Setter/getter

In this pattern, the UI component does not observe a Property. Instead, its API provides set/get functions for mutating its state.  This approach does not necessitate implementing `dispose`, and does not require a dedicated Property to control state.

Typical implementation:

```js
/ @constructor
function Component( …, options ) {
   
   options = _.extend( {
     enabled: true
   }, options );

   this._enabled = options.enabled; // @private
   this.updateEnabled( this._enabled );
}

// prototype properties
return inherit( Supertype, Component, {
  
  // @private
  updateEnabled: function( enabled ) {
      // enable or disable the component
  },

  // @public
  setEnabled: function( enabled ) {
    if ( enabled !== this._enabled ) {
      this._enabled = enabled;
      this.updateEnabled( enabled );
    }
  }
  set enabled( value ) { this.setEnabled( value ); },

  // @public
  getEnabled: function() { return this._enabled; },
  get enabled() { return this.getEnabled(); }
} );
```

Usage 3.1:

```js
var component = new Component(…);
var componentEnabledProperty = new Property( true );
var componentEnabledObserver = componentEnabledProperty.linkAttribute( component, ‘enabled’ );
var checkBox = new CheckBox( label, componentEnabledProperty );

// cleanup
componentEnabledProperty.unlinkAttribute( componentEnabledObserver );
checkBox.dispose();
```

Usage 3.2:

```js
var component = new Component(…);

var runningProperty = new Property( true ); // is the sim running?
var runningObserver = function( running ) {
  component.enabled = !running; // disable component when sim is running
  // … and anything else that needs to change
};
runningProperty.link( runningObserver );

var checkBox = new CheckBox( label, runningProperty );

// cleanup
runningProperty.unlink( runningObserver ); 
checkBox.dispose();
```

**Comment by pixelzoom** [comment]
# Pattern 4: Constructor option + setter/getter

This pattern combines patterns 1 & 3, providing both an option and setter/getter.  While this may be convenient for some scenarios, it complicates the implementation by providing 2 ways to do the same thing (generally something to avoid in an API).

Typical implementation:

```js
// @constructor
function Component( …, options ) {

   options = _.extend( {
       enabledProperty: new Property( true ) // {Property.<boolean>}
   }, options );

   this.enabledProperty = options.enabledProperty; // @public

   var enabledObserver = function( enabled ) {
     // enable or disable the component
   };
   this.enabledProperty.link( enabledObserver );

   // @private
   this.disposeComponent = function() {
     enabledProperty.unlink( enabledObserver );
   }
}

// prototype properties
return inherit( Supertype, Component, {

  // @public
  dispose: function() { this.disposeComponent(); }

 // @public
  setEnabled: function( enabled ) {
    if ( enabled !== this._enabled ) {
      this._enabled = enabled;
      this.updateEnabled( enabled );
    }
  }
  set enabled( value ) { this.setEnabled( value ); },

  // @public
  getEnabled: function() { return this._enabled; },
  get enabled() { return this.getEnabled(); } 
} );
```

Usage 4.1:

```js
var componentEnabledProperty = new Property( true );
var component = new Component( …, {
  enabledProperty: componentEnabledProperty
} );
var checkBox = new CheckBox( label, componentEnabledProperty );

// cleanup
component.dispose();
checkBox.dispose();
```

Usage 4.2 - use Property, ignore setter

```js
var componentEnabledProperty = new Property( true );
var component = new Component( …, {
  enabledProperty: componentEnabledProperty
} );

var runningProperty = new Property( true ); // is the sim running?
var checkBox = new CheckBox( label, runningProperty );

var runningObserver = function( running ) {
  componentEnabledProperty.set( !running );  // disable component when sim is running
  // … and anything else that needs to change
};
runningProperty.link( runningObserver );

// cleanup
component.dispose();
checkBox.dispose();
runningProperty.unlink( runningObserver ); 
```

Usage 4.3 - use setter, ignore Property:

```js
var component = new Component( … );

var runningProperty = new Property( true ); // is the sim running?
var checkBox = new CheckBox( label, runningProperty );

var runningObserver = function( running ) {
  component.enabled = !running;  // disable component when sim is running
  // … and anything else that needs to change
};
runningProperty.link( runningObserver );

// cleanup
component.dispose();
checkBox.dispose();
runningProperty.unlink( runningObserver ); 
```

**Comment by pixelzoom** [comment]
7/21/16 dev meeting: Consensus is to go with Pattern 4 (https://github.com/phetsims/sun/issues/241#issuecomment-234031982), and use Property or setter (but not both) in a situation, as appropriate.  Investigate factoring out setter/getter and enabledProperty into something like a mixin.

**Comment by samreid** [comment]
Investigate mixin *or* parent class *or* auxiliary library code.

**Comment by pixelzoom** [comment]
# Pattern 5: Component provide the Property

Noticed during dev meeting.... sun buttons actually create the `enabledProperty` (see ButtonModel) and don't allow you to provide one.  Decided that we should avoid this pattern.

Typical implementation:

```js
// @constructor
function Component( … ) {

  options = _.extend( {
    enabled: true
  }, option );

  // @public
   this.enabledProperty = new Property( options.enabled );
   }, options );
  this.enabledProperty.link( function( enabled ) {
    // enable or disable the component
  } );
}
```

Usage 5.1:

```js
var componentEnabledProperty = new Property( true );
var component = new Component( … );
var checkBox = new CheckBox( label, component.enabledProperty );

// cleanup
checkBox.dispose();
```

**Comment by pixelzoom** [comment]
Next task would be to change all existing UI components to implement Pattern 4.

**Comment by pixelzoom** [comment]
The consensus was Pattern 4, see https://github.com/phetsims/sun/issues/241#issuecomment-234322865.  And it looks like UI components have generally been converted to use that pattern.   So can this issue be closed?

A remaining problem is that we're add boilerplate code into each UI component individually, and then the boilerplate wasn't changed everywhere to meet PhET-iO requirements (see #257).

**Comment by pixelzoom** [comment]
From https://github.com/phetsims/sun/issues/241#issue-166639295:
> • What are the current patterns?
> • Are there other patterns that should be considered?

See options described above.

> • Should PhET standardize on one pattern, or is variation justified?

In https://github.com/phetsims/sun/issues/257 we decided to use the Mixin pattern, and it was implemented in `EnabledComponent`. 

> • Is this topic applicable only to UI components, or does it apply to other things (e.g. model elements) that can be enabled/disabled?

`EnabledComponent` is not specific to UI components, and can be used for any component that requires a dynamic "enabled" feature.  In https://github.com/phetsims/sun/issues/585 we decided to convert PhET UI components to use `EnabledComponent`.

Closing.
